### DISCUSSION TITLE
Should fzf shell integration be initialised before or after calling compinit ?

### DISCUSSION THREAD

**Discussion by xd003** [junegunn/fzf#3981] [Q&A] [answered]
Just wondering if the fzf shell integration for zsh
```
# Set up fzf key bindings and fuzzy completion
source <(fzf --zsh)
```
Should come before or after calling compinit ?

**Comment by LangLangBart** [comment, marked as answer]
`fzf` doesn't rely on the default zsh's completion system that is initialised when you call `compinit`[^1][^2]. Therefore, placing it before or after `compinit` should not matter.

**Example**: Setting up `fzf` completion without calling `autoload -Uz compinit && compinit` results in no issues.
```sh
# minimal zsh environment
command env -i HOME=$HOME TERM=$TERM USER=$USER LANG=en_US.UTF-8 PATH=$PATH /usr/local/bin/zsh -f
source <(fzf --zsh)
# the FZF_COMPLETION_TRIGGER is by default set to '**'
ls **<TAB>
```

However, in the zsh setup files I've observed, it is placed after `autoload -U compinit && compinit`, not because it's necessary, but to ensure default user configurations are initialized first. This allows third-party tools to override them if needed, giving the setup a logical structure.

Other tools, for example, `lincheney/fzf-tab-completion`[^3] must be called after `compinit` to work properly.

**Example**: Setting up `lincheney/fzf-tab-completion` without calling `autoload -Uz compinit && compinit`
```sh
command env -i HOME=$HOME TERM=$TERM USER=$USER LANG=en_US.UTF-8 PATH=$PATH /usr/local/bin/zsh -f
source /Users/paria/Developer/fzf-tab-completion/zsh/fzf-zsh-completion.sh
bindkey '^I' fzf_completion
ls <TAB>
# _fzf_completion:93: command not found: _main_complete

autoload -Uz compinit && compinit
ls <TAB>
```

---

Did you encounter any issues that led you to ask this question?

[^1]: [zsh source code: /Completion/compinit](https://sourceforge.net/p/zsh/code/ci/master/tree/Completion/compinit)
[^2]: [A User's Guide to the Z-Shell: Chapter 6: Completion, old and new](https://zsh.sourceforge.io/Guide/zshguide06.html#l144)
[^3]: [GitHub - lincheney/fzf-tab-completion: Tab completion using fzf](https://github.com/lincheney/fzf-tab-completion)

**Reply by xd003** [reply]
Thank you for your detailed reply. I didn't encounter any issues per se, but I was curious about the best practices, as the README doesn't cover them.
